### PR TITLE
Remoting rc3

### PR DESCRIPTION
**Pull request by aalmiray**
Fix for #155

---
This change is [Reviewable](https://reviewable.io/reviews/canoo/dolphin-platform/161)

### CODE REVIEW

File: remoting/dolphin-remoting-shared/src/main/groovy/org/opendolphin/StringUtil.groovy [hunk 1]
```diff
@@ -28,6 +27,15 @@ class StringUtil {
      *         blank.
      */
     public static boolean isBlank(String str) {
-        return str == null || str.trim().length() == 0
+        if (str == null || str.length() == 0) {
```
Comment (netopyr) [inline remoting/dolphin-remoting-shared/src/main/groovy/org/opendolphin/StringUtil.groovy]: What was broken in the old implementation?

Comment (aalmiray) [inline remoting/dolphin-remoting-shared/src/main/groovy/org/opendolphin/StringUtil.groovy]: `str.trim().length()` is a bad way to check if a `String` contains only whitespaces.
Given the following inputs

 * null
 * ''
 * '     ' (5 spaces)
 * '    c' (4 spaces to left)
 * 'c    ' (4 spaces to the right)
 * '  c  ' (2 spaces each side)

The previous version would have resulted in a new `String` if `str.length > 0` and if there where whitespace chars at either end. Trim also checks against the white space char only, disregarding tabs, returns, and other platform specific whitespace chars. Finally, trim would have touched the last string on both sides, where as the newer `isBlank` stops at char[2] because it's not whitespace.

In other words, `isBlank` is better because
 - does not create new `String` instances
 - checks whitespace agains `Character.isWhiteSpace()`
 - does not iterate the whole char[]

Comment (netopyr) [inline remoting/dolphin-remoting-shared/src/main/groovy/org/opendolphin/StringUtil.groovy]: I can understand the isWhiteSpace() argument, though I wonder if it really applies, because trim() does regard tabs, returns etc. and the only difference are rather exotic whitespaces.

Performance-wise I doubt that the new implementation is faster. I would assume that most of the time the method will be called with an empty String, null, and Strings with no whitespaces on either end. In the last case, the new implementation is probably slower, because it copies the whole char array in any case, while the old implementation skips pretty much all steps.

Comment (aalmiray) [inline remoting/dolphin-remoting-shared/src/main/groovy/org/opendolphin/StringUtil.groovy]: I also thought `trim()` would take care of whitespace chars, sadly it does not. Here's the code copied directly form the JSL

    public String trim() {
        int len = value.length;
        int st = 0;
        char[] val = value;    /* avoid getfield opcode */

        while ((st < len) && (val[st] <= ' ')) {
            st++;
        }
        while ((st < len) && (val[len - 1] <= ' ')) {
            len--;
        }
        return ((st > 0) || (len < value.length)) ? substring(st, len) : this;
    }

The new implementation does not copy any chars, it's the old one that does it (because that's what `trim()` does).
Performance wise the new impl is faster than the old one in the last two examples. When the input  `String` contains all whitespace chars then they have the same speed, but the old one requires more memory (new instance) and only checks a single whitespace char.

**Comment by hendrikebbers** [comment]
Spring & JavaEE is working for me, Kumulz still not working. 
LGTM

File: remoting/dolphin-remoting-shared/src/main/groovy/org/opendolphin/StringUtil.groovy (hunk 1 above)
Comment (netopyr) [inline remoting/dolphin-remoting-shared/src/main/groovy/org/opendolphin/StringUtil.groovy]: val[st] <= ' ' takes care of all chars that are smaller than space, which also includes tab, cr, lf, and others.

In the loop you call String.toCharArray(), which copies the char array.

trim() creates a new String only if necessary. If you look into substring(), you will notice that it returns the original String if beginIndex is 0 and endIndex equals the length of the array, i.e. in the last one of our cases we do not create or copy anything in trim().

Comment (Dierk) [inline remoting/dolphin-remoting-shared/src/main/groovy/org/opendolphin/StringUtil.groovy]: The first rule of performance optimization is ...

**Comment by aalmiray** [comment]
What kind of error do you get when running the Kumuluz server? Might be worth creating a ticket for it

**Comment by hendrikebbers** [comment]
See issue https://github.com/canoo/dolphin-platform/issues/162

**Comment by hendrikebbers** [comment]
@aalmiray Dependecy 'javax.inject:javax.inject:1' in remoting server can be removed. I think it's the best to do this directly in this pull request.

**Comment by aalmiray** [comment]
why? all consumers either use Spring or CDI, right? If so then this dependency should be superfluous, but it isn't. Removing it from `dolphin-remoting-server` makes `todo-server` fail compilation. this is because `javax.inject` is not explicitly added on this project, nor does it come from `dolphin-platform-server`.

What I mean is, sure we can remove it from `dolphin-remoting-server` but then it must be added to `dolphin-platform-server`, other wise you can't build apps like todo-example without explicitly adding `javax.inject` first.

**Comment by hendrikebbers** [comment]
@aalmiray About @Inject: I think this is not right. Spring uses @Autowire internally and JavaEE uses @Inject. You can use @Inject in your Spring application if you want. In that case you need to add the CDI-JSR dependency. This is based on some Spring magic. I think adding it by default is just wrong. And yes, we need to add it to our Todo-Server that would be the right place. In the remoting layer @Inject is never used.

**Comment by hendrikebbers** [comment]
added #170

**Comment by aalmiray** [comment]
Spring can use `@Inject` without and CDI dependency, you only need the JSR-330 JAR in the classpath. that's because JSR-330 is the lowest common denominator for performing DI. The `javax.inject` dependency was adding to the remoting layer many months ago in order to provide DI compatible classes but the code was never added.

It's my intention to add those classes in a future PR. Remember, the presence this dependency and the annotations on classes *do not* force consumers of the remoting layer to *always* rely on DI. As a _very_ positive side effect of thinking in DI terms when designing APIs is that we can make them more testable and avoid circular dependencies (which we already have in Dolphin, ModelStore, and Connector).

**Comment by hendrikebbers** [comment]
Ok, based on the changes that you want to do for the server-remoting I understand why you want this dependency. 
I still think that it's not good to use DI on that layer. The lib must work with Spring and JavaEE that means @Inject will be handled by the used enterprise framework. The first problem here is, that both have different ways to define scopes. I think providing (C)DI support for application development is a must but using it in internal APIs might create some problems (if you don't use 1 fix provider)

**Comment by aalmiray** [comment]
I guess I'm missing something here as I don't see why `@Inject` would cause trouble with scopes. As far as I understand it, Spring treats `@Inject` as an alias for `@Autowired`. CDI on the other hand relies on `@Inject` to perform injection.

`@Autowired` does not declare a scope, nor does `@Inject`. You must use additional annotations to get the desired scope. Those scope annotations _may_ be compatible with JSR-330 `Scope` but not necessarily. The intention for adding `javax.inject` as a dependency is to design and annotate constructors with `@Inject` where we know dependencies must be final; for other cases I suppose we can use field injection. None of these require a particular scope to be used. That's up to the next layer (DP) or the application to decide.

**Comment by hendrikebbers** [comment]
Let's say you want to define the `ServerDolphin` as injectable. By doing so you need something that provides an injectable instance of the `ServerDolphin`. In Spring you can create such an provider by using the `@Bean` annotation. In JavaEE you need the `@Provides` annotation. If you define no scope for a bean type JavaEE defines the dependent scope as default. I don't know what Spring does in this case.

**Comment by aalmiray** [comment]
Spring would put the bean in singleton scope by default. You can change this behavior by either
- annotating with `@Scope` the method that produces the bean, i.e, the one that is annotated with `@Bean`.
- annotating the class itself with `@Scope`. This definitely is not what I intend to happen in this API.

The other two methods which Spring uses to locate components are annotating the class with either `@Component` or `@Service`. Also not happening for the remoting APIs.

As an aside, Guice is capable of providing dynamic bindings (that is, automatically create a binding) when it finds a type that fits the bill, even if no explicit binding was declared.

So.

Annotating constructors fields with `@Inject` only makes the classes ready for DI, but do not force a DI container to be used. Also it does not force the hand for a particular DI solution, so neither Spring nor CDI should suffer from this.

**Comment by hendrikebbers** [comment]
> Annotating constructors fields with @Inject only makes the classes ready for DI, but do not force a DI container to be used. Also it does not force the hand for a particular DI solution, so neither Spring nor CDI should suffer from this.

Right! So you don't want to use DI in the remoting layer but prepare the main classes for DI by adding constructors that support DI. In the remoting layer you will still call the constructors directly in code. Is that your plan?

**Comment by aalmiray** [comment]
Correct. The production API of the remoting layer will not expose nor require a specific container. It'll only make sure the API is DI ready. Tests of the remoting layer may use Guice, Spring or some other container; these dependencies will not be exposed to DP nor any other module.

**Comment by hendrikebbers** [comment]
Ok, now I understand what you want to do :)
